src/python/pants/engine/selectors.py

```diff
@@ -154,7 +154,8 @@ def __repr__(self):
                                        type_or_constraint_repr(self.product),
                                        self.projected_subject.__name__,
                                        repr(self.fields),
-                                       self.input_product.__name__)
+                                       self.input_product.__name__ if hasattr(
+                                         self.input_product, '__name__') else repr(self.input_product))
```

Comment:
Is it better to do
getattr(self.input_product, '__name__', repr(self.input_product))